### PR TITLE
added PAXG collateral with GemJoin9

### PR DESCRIPTION
**Pull request by hexonaut**
# Description

Collateral Type (e.g. `ETH-A`): PAXG-A

COB Team Name or Author(s): @hexonaut @godsflaw 

GemJoin Adapter: `GemJoin9`

Is this GemJoin adapter new? : Yes

If so, why is a new one needed? : PAXG has an optional transaction fee which needs to be deducted from the joined gem amount.

# Contribution Checklist

- [x] Collateral token contract has been added in `src/tokens/` as `TOKEN.sol`, if not already present
- [x] If needed, a new `GemJoin` has been added in `src/` as `join-X.sol`
- [x] A new test (`testGemJoinX_TOKEN()`) has been added to `src/join.t.sol` and follows previous tests as examples
- [x] If needed, other tests have been added to verify functionality of the new `GemJoin`:
    - [x] `testFailGemJoinXJoinWad()`
    - [x] `testFailGemJoinXExitWad()`
    - [x] `testFailGemJoinXJoin()`
    - [x] `testFailGemJoinXExit()`
    - [x] `testFailJoinAfterCageGemJoinX()`
- [x] README has been updated to reflect a new token and `GemJoin` addition

# Checklist

- [x] Confirm that token contract matches the verified source on etherscan mainnet
- [x] If a new `GemJoin` is added, ensure the above description is accurate
- [x] If a new `GemJoin` is added, ensure a new `GemJoin` has been added with the correct name
- [x] Ensure README is updated
- [x] Manually review the new test(s)
- [x] Manually run dapp tests in `src/join.t.sol` to confirm they all pass

### CODE REVIEW

File: src/join-9.sol
Comment (lucas-manuel) [inline src/join-9.sol:72]: I think itd be worth adding a `uint256 beforeBal = gem.balanceOf(address(this));` before the `transferFrom` and then an `assert`: 
`assert(sub(gem.balanceOf(address(this), beforeBal) == sub(wad, gem.getFeeFor(wad))));` to avoid any surprises down the road. Especially since this is upgradeable. cc @godsflaw

**Review by lucas-manuel — CHANGES_REQUESTED** [review]
LGTM, left one comment.

File: src/join-9.sol
Comment (hexonaut) [reply src/join-9.sol:72]: Makes sense to me.

**Review by lucas-manuel — APPROVED** [review]
LGTM

File: src/join-9.sol
Comment (brianmcmichael) [reply src/join-9.sol:75]: Was going to come back and mention this after looking it over yesterday. Glad this got added. LGTM.

File: src/join-9.sol
Comment (godsflaw) [inline src/join-9.sol:71]: this variable should be called `sad`

File: src/join-9.sol
Comment (godsflaw) [inline src/join-9.sol:74]: We can call this `bal`.  It makes it more conformant to the Tether adapter.

**Review by godsflaw — CHANGES_REQUESTED** [review]
I made a few nits about variable names in the join.  Good work on accounting for fees, and thanks for testing that edge case.  I will flag here that we need to check upstream of this just to make sure we don't need changes in `dss-proxy-actions`.

**Review by lucas-manuel — APPROVED** [review]

**Review by brianmcmichael — DISMISSED** [review]

**Comment by hexonaut** [comment]
Switching over to a wrapped gem implementation to avoid the double transfer fees in dss-proxy-actions. Will add a PR for dss-proxy-actions too as this needs new functionality.

**Comment by hexonaut** [comment]
Switched back as I figured out a way to do a 1 transfer join without a wrapper.

**Comment by gbalabasquer** [comment]
I see the advantages of doing it in this way, however this restricts the usage to only proxy users.
Do we really want to do so for a collateral type?
In case of other more specifics modules I would probably be fine with it, but here we are talking about a collateral type unique adapter which will only be able to be used via a proxy.
I feel more confortable with a wrapper and the standard `GemJoin` and not leaving anyone behind.

**Comment by gbalabasquer** [comment]
Btw sorry if this has been already discussed and I came a bit late to the discussion, I just looked at this for the first time today.

**Comment by hexonaut** [comment]
Regular users can still interact with the join(usr,wad) overload.

**Comment by gbalabasquer** [comment]
Ah sorry I focused in that `join` function that I didn't see the other one.

File: src/join-9.sol
Comment (gbalabasquer) [inline src/join-9.sol:97]: I think a better option would be making `join(address usr)` `public` and then having:
```
function join(address usr, uint256 wad) external note {
     require(gem.transferFrom(msg.sender, address(this), wad), "GemJoin/failed-transfer");
     join(usr);
}
```
If I'm not wrong this would make us save two external call and two `SLOAD`s with the caveat that now you will be joining an `amount >= wad` (if by any chance there was balance not recorded in the adapter). However that would be basically the same behaviour than when we are using the other `join` function.

File: src/join-9.sol
Comment (gbalabasquer) [inline src/join-9.sol:108]: nit: missing breakline

File: src/join.t.sol
Comment (gbalabasquer) [inline src/join.t.sol:977]: nit: missing breakline

File: src/join-9.sol
Comment (gbalabasquer) [inline src/join-9.sol:79]: `total` can be assigned to a memory variable `_total` in order to reuse it in line 83 and save 1 `SLOAD`.

File: src/join-9.sol
Comment (hexonaut) [reply src/join-9.sol:97]: Ah yes I like this better. Since the fee is coming out, the assumption is already broken that `wad` = the number of gems you will get, so might as well simplify. This also generalizes the adapter to any token that collects fees, and removes the requirement of `gem.getFeeFor(wad)`. Good call.

File: src/join-9.sol
Comment (brianmcmichael) [inline src/join-9.sol:87]: The `note` event is emitted twice here. Open for debate whether we just want to remove the `note` from this function, or take most of the logic in `join(usr)` and add it to an internal function, which gets shared by `join(usr)` and `join(usr,wad)`. If we did that we could make `join(usr)` into an `external` function.

Comment (hexonaut) [reply src/join-9.sol:87]: What are the repercussion of just removing the `note`?

Comment (gbalabasquer) [reply src/join-9.sol:87]: We should just remove the `note` modifier from `join(address usr, uint256 wad)`, that was actually my example in the comment above ^^ (should have made it explicit though).

File: src/join-9.sol
Comment (brianmcmichael) [inline src/join-9.sol:3]: nit: `join-9.sol`

File: src/tokens/PAXG.sol
Comment (brianmcmichael) [inline src/tokens/PAXG.sol:44]: Maybe we should send them to the faucet.

File: src/tokens/PAXG.sol
Comment (brianmcmichael) [inline src/tokens/PAXG.sol:76]: nit: spelling, I think `principal` here

Comment (hexonaut) [reply src/tokens/PAXG.sol:76]: Looks like they misspelled in the actual token code.

**Review by gbalabasquer — APPROVED** [review]
LGTM

File: src/join-9.sol
Comment (brianmcmichael) [inline src/join-9.sol:87]: This could probably use a detailed comment that this function can be used without the proxy-actions, but will double charge.

Comment (hexonaut) [reply src/join-9.sol:87]: Done.

File: src/join-9.sol
Comment (naszam) [inline src/join-9.sol:6]: we might want to update the license here
```suggestion
// Copyright (C) 2020-2022 Dai Foundation
```

File: src/join-9.sol
Comment (naszam) [inline src/join-9.sol:23]: I think this can be removed in favour of events
```suggestion
```

File: src/join-9.sol
Comment (naszam) [inline src/join-9.sol:46]:
```suggestion
    // --- Auth ---
    mapping (address => uint256) public wards;
    function rely(address usr) external auth {
        wards[usr] = 1;
        emit Rely(usr);
    }
    function deny(address usr) external auth {
        wards[usr] = 0;
        emit Deny(usr);
    }
    modifier auth {
        require(wards[msg.sender] == 1, "GemJoin9/not-authorized");
        _;
    }
```

**Comment by hexonaut** [comment]
@naszam can you take this PR over?

File: src/join.t.sol
Comment (naszam) [inline src/join.t.sol:927]:
```suggestion
        PAXG paxg = new PAXG(100 * 10 ** 18, 0);
```

File: src/join.t.sol
Comment (naszam) [inline src/join.t.sol:841]:
```suggestion
        assertEq(paxgJoin.total(), paxg.balanceOf(address(paxgJoin)));
    }
 
```

**Review by godsflaw — APPROVED** [review]
Approving so this can be merged, but this needs another review once it's re-opened given how much time has passed.

File: src/join-9.sol
Comment (naszam) [reply src/join-9.sol:6]: addressed in #33

**Pull request merged**
